JvmHinter does not work for generic types on OpenJ9-JVM

- [x] I am running the latest version
- [x] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed

### Expected Behavior

Mocks of generic return types share same behavior cross-jvm implementation

### Current Behavior

OpenJ9 causes MockK to throw a Class-Cast Exception when mocking a function with generic return-type.

### Failure Information (for bugs)

The Auto-Hinter seems to not work in certain cases on the OpenJ9 JVM Implementation - on Hotspot it works fine.

#### Steps to Reproduce

I already have an open issue here: https://github.com/mockk/mockk/issues/717, so I extended the Repo with the code to reproduce (can be found at https://github.com/Staffbase/mockK-bug) by the [`GenericHinterTest`](https://github.com/Staffbase/mockK-bug/blob/main/src/test/java/com/staffbase/bugs/GenericHinterTest.kt) - it will fail on OpenJ9)

#### Context

Please provide any relevant information about your setup. This is important in case the issue is not reproducible except for under certain conditions.

* MockK version: 1.12.0
* OS: Dockers `ubuntu-latest`, Arch Linux x86_64 (Kernel: 5.14.8-arch1-1), macOS
* Kotlin version: 1.5.31
* JDK version: openjdk 16.0.1 - OpenJ9 0.26.0 (adoptopenjdk/openjdk16-openj9:alpine-slim)
* JUnit version: 5.8.1
* Type of test: Unit-Test


#### Failure Logs

```
Oct 17, 2021 4:46:17 PM io.mockk.impl.log.JULLogger warn
WARNING: Non instrumentable classes(skipped): class java.lang.Object
```

#### Stack trace

```
io.mockk.MockKException: Class cast exception happened.
Probably type information was erased.
In this case use `hint` before call to specify exact return type of a method.

	at app//io.mockk.impl.InternalPlatform.prettifyRecordingException(InternalPlatform.kt:73)
	at app//io.mockk.impl.eval.RecordedBlockEvaluator.record(RecordedBlockEvaluator.kt:66)
	at app//io.mockk.impl.eval.EveryBlockEvaluator.every(EveryBlockEvaluator.kt:30)
	at app//io.mockk.MockKDsl.internalEvery(API.kt:92)
	at app//io.mockk.MockKKt.every(MockK.kt:98)
	at app//com.staffbase.bugs.GenericHinterTest.auto hints do not work(GenericHinterTest.kt:19)
	at java.base@16.0.2/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base@16.0.2/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base@16.0.2/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@16.0.2/java.lang.reflect.Method.invoke(Method.java:567)
	at app//org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at app//org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at app//org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at app//org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at app//org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor$$Lambda$173/0x00000000e0578080.apply(Unknown Source)
	at app//org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at app//org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall$$Lambda$174/0x00000000e05788f0.apply(Unknown Source)
	at app//org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at app//org.junit.jupiter.engine.execution.ExecutableInvoker$$Lambda$330/0x00000000e05b6cc8.apply(Unknown Source)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at app//org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at app//org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor$$Lambda$342/0x00000000e05ba580.execute(Unknown Source)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$268/0x00000000e05a8f88.execute(Unknown Source)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$267/0x00000000e05a8dd0.invoke(Unknown Source)
	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$266/0x00000000e05a8a90.execute(Unknown Source)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService$$Lambda$272/0x00000000e05aa600.accept(Unknown Source)
	at java.base@16.0.2/java.util.ArrayList.forEach(ArrayList.java:1511)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$268/0x00000000e05a8f88.execute(Unknown Source)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$267/0x00000000e05a8dd0.invoke(Unknown Source)
	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$266/0x00000000e05a8a90.execute(Unknown Source)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService$$Lambda$272/0x00000000e05aa600.accept(Unknown Source)
	at java.base@16.0.2/java.util.ArrayList.forEach(ArrayList.java:1511)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$268/0x00000000e05a8f88.execute(Unknown Source)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$267/0x00000000e05a8dd0.invoke(Unknown Source)
	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$266/0x00000000e05a8a90.execute(Unknown Source)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at app//org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at app//org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator$$Lambda$228/0x000000008eff2610.accept(Unknown Source)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base@16.0.2/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base@16.0.2/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base@16.0.2/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@16.0.2/java.lang.reflect.Method.invoke(Method.java:567)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at jdk.proxy1/jdk.proxy1.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: java.lang.ClassCastException: java.lang.Object incompatible with java.lang.String
	at app//com.staffbase.bugs.GenericHinterTest$auto hints do not work$1.invoke(GenericHinterTest.kt:20)
	at app//com.staffbase.bugs.GenericHinterTest$auto hints do not work$1.invoke(GenericHinterTest.kt:19)
	at app//io.mockk.impl.eval.RecordedBlockEvaluator$record$block$1.invoke(RecordedBlockEvaluator.kt:25)
	at app//io.mockk.impl.eval.RecordedBlockEvaluator$enhanceWithRethrow$1.invoke(RecordedBlockEvaluator.kt:78)
	at app//io.mockk.impl.recording.JvmAutoHinter.autoHint(JvmAutoHinter.kt:23)
	at app//io.mockk.impl.eval.RecordedBlockEvaluator.record(RecordedBlockEvaluator.kt:40)
	... 103 more
```

#### Minimal reproducible code (the gist of this issue)

```kotlin
// -----------------------[ GRADLE DEFINITIONS ] -----------------------
plugins {
    java
    kotlin("jvm") version "1.5.31"
}

group = "com.staffbase"
version = "1.0-SNAPSHOT"

repositories {
    mavenCentral()
}

java {
    toolchain {
        languageVersion.set(JavaLanguageVersion.of(16))
        implementation.set(JvmImplementation.J9)
    }
}

tasks.withType<Test> {
    useJUnitPlatform()
}

dependencies {
    implementation(kotlin("stdlib"))

    testImplementation("io.mockk:mockk:1.12.0")
    testImplementation("org.junit.jupiter:junit-jupiter-api:5.8.1")
    testImplementation("org.junit.jupiter:junit-jupiter-engine:5.8.1")
    testImplementation("org.junit.jupiter:junit-jupiter-params:5.8.1")
}



// -----------------------[ HELPER CLASS - Provider.java ] -----------------------
package com.staffbase.entities;

public interface Provider<T> {
    T get();
}



// -----------------------[ YOUR CODE STARTS HERE ] -----------------------
package com.staffbase.bugs

import com.staffbase.entities.Provider
import io.mockk.every
import io.mockk.mockk
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

typealias StringProviderType = Provider<String>

class GenericHinterTest {

    @Test
    fun `auto hints do not work`() {

        val expected = "Hello, World!"
        val mockedProvider = mockk<Provider<String>>()

        every {
            mockedProvider.get()
        } returns expected

        Assertions.assertEquals(expected, mockedProvider.get())
    }

    @Test
    fun `Typealias fails as well`() {
        val expected = "Hello, World!"

        val mockedProvider = mockk<StringProviderType>()

        every {
            mockedProvider.get()
        } returns expected

        Assertions.assertEquals(expected, mockedProvider.get())
    }

    @Test
    fun `manually setting the hints works`() {

        val expected = "Hello, World!"
        val mockedProvider = mockk<Provider<String>>()

        every {
            mockedProvider.hint(String::class).get()
        } returns expected

        Assertions.assertEquals(expected, mockedProvider.get())

        every {
            hint(String::class)
            mockedProvider.get()
        } returns expected

        Assertions.assertEquals(expected, mockedProvider.get())
    }

    @Test
    fun `class typing the provider works as well`() {

        val expected = "Hello, World!"

        class StringProvider : Provider<String> {
            override fun get() = "something"
        }

        val mockedProvider = mockk<StringProvider>()

        every {
            mockedProvider.get()
        } returns expected

        Assertions.assertEquals(expected, mockedProvider.get())
    }
}
// -----------------------[ YOUR CODE ENDS HERE ] -----------------------
```

--- a/modules/mockk/src/jvmMain/kotlin/io/mockk/impl/recording/JvmAutoHinter.kt
+++ b/modules/mockk/src/jvmMain/kotlin/io/mockk/impl/recording/JvmAutoHinter.kt
@@ -50,11 +50,14 @@ class JvmAutoHinter : AutoHinter() {
         } ?: throw ex
 
     companion object {
-        // JDK 8: net.bytebuddy.renamed.java.lang.Object$ByteBuddy$As29nsJf$ByteBuddy$877l7O7D cannot be cast to io.mockk.impl.recording.states.CallRecordingState
-        // JDK 9:
-        // JDK 10:
-        // JDK 11: class net.bytebuddy.renamed.java.lang.Object$ByteBuddy$rpycQEYo$ByteBuddy$bHEk1ADY cannot be cast to class java.lang.String (net.bytebuddy.renamed.java.lang.Object$ByteBuddy$rpycQEYo$ByteBuddy$bHEk1ADY is in unnamed module of loader net.bytebuddy.dynamic.loading.ByteArrayClassLoader @19569ebd; java.lang.String is in module java.base of loader 'bootstrap')
-        val exceptionMessage = Regex("cannot be cast to (class )?(.+/)?(.+?)( \\((.+)\\))?$")
+        // HotSpot pre-JEP358 (JDK 8-13):
+        //   net.bytebuddy.renamed.java.lang.Object$ByteBuddy$As29nsJf$ByteBuddy$877l7O7D cannot be cast to io.mockk.impl.recording.states.CallRecordingState
+        // HotSpot JEP358 (JDK 14+):
+        //   class net.bytebuddy.renamed.java.lang.Object$ByteBuddy$rpycQEYo$ByteBuddy$bHEk1ADY cannot be cast to class java.lang.String (net.bytebuddy.renamed.java.lang.Object$ByteBuddy$rpycQEYo$ByteBuddy$bHEk1ADY is in unnamed module of loader net.bytebuddy.dynamic.loading.ByteArrayClassLoader @19569ebd; java.lang.String is in module java.base of loader 'bootstrap')
+        // OpenJ9 / IBM Semeru (all releases):
+        //   net.bytebuddy.renamed.java.lang.Object$ByteBuddy$As29nsJf$ByteBuddy$877l7O7D incompatible with java.lang.String
+        val exceptionMessage =
+            Regex("(?:cannot be cast to|incompatible with) (class )?(.+/)?(.+?)( \\((.+)\\))?$")
 
         val log = Logger<JvmAutoHinter>()
     }

--- a/modules/mockk/src/jvmTest/kotlin/io/mockk/impl/recording/JvmAutoHinterTest.kt
+++ b/modules/mockk/src/jvmTest/kotlin/io/mockk/impl/recording/JvmAutoHinterTest.kt
@@ -1,0 +1,67 @@
+package io.mockk.impl.recording
+
+import io.mockk.impl.recording.JvmAutoHinter.Companion.exceptionMessage
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Assertions.assertNull
+import org.junit.jupiter.api.Test
+
+class JvmAutoHinterTest {
+    @Test
+    fun extractsTargetClassFromHotSpotPreJep358Message() {
+        val msg =
+            "net.bytebuddy.renamed.java.lang.Object\$ByteBuddy\$abc " +
+                "cannot be cast to java.lang.String"
+        assertEquals(
+            "java.lang.String",
+            exceptionMessage
+                .find(msg)
+                ?.groups
+                ?.get(3)
+                ?.value,
+        )
+    }
+
+    @Test
+    fun extractsTargetClassFromHotSpotJep358Message() {
+        val msg =
+            "class net.bytebuddy.renamed.java.lang.Object\$ByteBuddy\$abc " +
+                "cannot be cast to class java.lang.String " +
+                "(net.bytebuddy.renamed.java.lang.Object\$ByteBuddy\$abc is in unnamed module " +
+                "of loader net.bytebuddy.dynamic.loading.ByteArrayClassLoader @19569ebd; " +
+                "java.lang.String is in module java.base of loader 'bootstrap')"
+        assertEquals(
+            "java.lang.String",
+            exceptionMessage
+                .find(msg)
+                ?.groups
+                ?.get(3)
+                ?.value,
+        )
+    }
+
+    @Test
+    fun extractsTargetClassFromOpenJ9Message() {
+        val msg =
+            "net.bytebuddy.renamed.java.lang.Object\$ByteBuddy\$abc " +
+                "incompatible with java.lang.String"
+        assertEquals(
+            "java.lang.String",
+            exceptionMessage
+                .find(msg)
+                ?.groups
+                ?.get(3)
+                ?.value,
+        )
+    }
+
+    @Test
+    fun returnsNullOnUnrecognisedMessage() {
+        assertNull(
+            exceptionMessage
+                .find("some other failure")
+                ?.groups
+                ?.get(3)
+                ?.value,
+        )
+    }
+}

